be/src/exec/pipeline/exchange/mem_limited_chunk_queue.cpp

```diff
@@ -0,0 +1,528 @@
+}
+
+Status MemLimitedChunkQueue::push(const ChunkPtr& chunk) {
+    RETURN_IF_ERROR(_get_io_task_status());
+    std::unique_lock l(_mutex);
+    if (UNLIKELY(_chunk_builder == nullptr)) {
+        _chunk_builder = chunk->clone_empty();
+    }
+    int32_t closed_source_number = _consumer_number - _opened_source_number;
+
+    DCHECK(_tail != nullptr) << "tail won't be null";
+    DCHECK(_tail->next == nullptr) << "tail should be the last block";
+
+    // create a new block
+    if (_tail->memory_usage >= _opts.block_size) {
+        Block* block = new Block();
+        block->next = nullptr;
+        _tail->next = block;
+        _tail = block;
+    }
+
+    _total_accumulated_rows += chunk->num_rows();
+    _total_accumulated_bytes += chunk->memory_usage();
+
+    Cell cell;
+    cell.chunk = chunk;
+    cell.used_count = closed_source_number;
+    cell.accumulated_rows = _total_accumulated_rows;
+    cell.accumulated_bytes = _total_accumulated_bytes;
+    _tail->cells.emplace_back(cell);
+    _tail->memory_usage += chunk->memory_usage();
+
+    size_t in_memory_rows = _total_accumulated_rows - _flushed_accumulated_rows + _current_load_rows;
+    size_t in_memory_bytes = _total_accumulated_bytes - _flushed_accumulated_bytes + _current_load_bytes;
+
+#ifndef BE_TEST
+    _peak_memory_rows_counter->set(in_memory_rows);
+    _peak_memory_bytes_counter->set(in_memory_bytes);
+#endif
+    return Status::OK();
+}
+
+bool MemLimitedChunkQueue::can_push() {
+    std::shared_lock l(_mutex);
+    size_t unconsumed_bytes = _total_accumulated_bytes - _fastest_accumulated_bytes;
+    // if the fastest consumer still has a lot of data to consume, it will no longer accept new input.
+    if (unconsumed_bytes >= _opts.max_unconsumed_bytes) {
+        return false;
+    }
+
+    size_t in_memory_bytes = _total_accumulated_bytes - _flushed_accumulated_bytes;
+    if (in_memory_bytes >= _opts.memory_limit && _next_flush_block->next != nullptr) {
+        if (bool expected = false; _has_flush_io_task.compare_exchange_strong(expected, true)) {
+            TEST_SYNC_POINT("MemLimitedChunkQueue::can_push::before_submit_flush_task");
+            auto status = _submit_flush_task();
+            if (!status.ok()) {
+                _update_io_task_status(status);
+                return true;
+            }
+        }
+        return false;
+    }
+
+    return true;
+}
+
+StatusOr<ChunkPtr> MemLimitedChunkQueue::pop(int32_t consumer_index) {
+    DCHECK(consumer_index <= _consumer_number);
+    RETURN_IF_ERROR(_get_io_task_status());
+    std::unique_lock l(_mutex);
+
+    DCHECK(_consumer_progress[consumer_index] != nullptr);
+    auto iter = _consumer_progress[consumer_index].get();
+    if (!iter->has_next()) {
+        if (_opened_sink_number == 0) {
+            return Status::EndOfFile("no more data");
+        }
+        return Status::InternalError("unreachable path");
+    }
+    iter->move_to_next();
+    DCHECK(iter->get_block()->in_mem);
```

Comment:
what will happen if flush thread flush this block when wait lock in 156?

Reply:
fixed